tt_metal/impl/dispatch/dispatch_core_manager.hpp

```diff
@@ -0,0 +1,136 @@
+            assignment.command_dispatcher = assignment.completion_queue_interface;
+        }
+        return assignment.completion_queue_interface.value();
+    }
+
+    const tt_cxy_pair &command_dispatcher_core(chip_id_t device_id, uint16_t channel, uint8_t cq_id) {
+        dispatch_core_types_t &assignment = this->dispatch_core_assignments[device_id][channel][cq_id];
+        if (assignment.command_dispatcher.has_value()) {
+            return assignment.command_dispatcher.value();
+        }
+        CoreCoord command_dispatcher_coord = this->get_next_available_dispatch_core(device_id);
+        chip_id_t mmio_device_id = tt::Cluster::instance().get_associated_mmio_device(device_id);
+        assignment.command_dispatcher = tt_cxy_pair(device_id, command_dispatcher_coord.x, command_dispatcher_coord.y);
+        std::cout << "For device " << device_id << " channel " << channel << " CQ " << std::to_string(cq_id) << " command dispatcher location: " << assignment.command_dispatcher.value().str() << std::endl;
+        if (mmio_device_id == device_id) {
+            // For MMIO devices completion queue core is same as command dispatcher core
+            TT_ASSERT(not assignment.completion_queue_interface.has_value(), "Command dispatcher core must match completion queue interface core for MMIO device {}", device_id);
+            assignment.completion_queue_interface = assignment.command_dispatcher;
+        }
+        return assignment.command_dispatcher.value();
+    }
+
+    const std::optional<tt_cxy_pair> &remote_processor_core(chip_id_t device_id, uint16_t channel, uint8_t cq_id) {
+        TT_THROW("Do not currently support programming remote processor dispatch core. See https://github.com/tenstorrent-metal/tt-metal/issues/3953");
+        return this->dispatch_core_assignments[device_id][channel][cq_id].remote_processor;
+    }
+
+    const std::optional<tt_cxy_pair> &remote_signaller_core(chip_id_t device_id, uint16_t channel, uint8_t cq_id) {
+        TT_THROW("Do not currently support programming remote signaller dispatch core. See https://github.com/tenstorrent-metal/tt-metal/issues/3954");
+        return this->dispatch_core_assignments[device_id][channel][cq_id].remote_signaller;
+    }
+
+   private:
+    dispatch_core_manager(uint8_t num_hw_cqs) {
+        for (chip_id_t device_id = 0; device_id < tt::Cluster::instance().number_of_devices(); device_id++) {
+            std::list<CoreCoord> &logical_dispatch_cores = this->available_dispatch_cores_by_device[device_id];
+            for (const CoreCoord &logical_dispatch_core : tt::get_logical_dispatch_cores(device_id, num_hw_cqs)) {
+                logical_dispatch_cores.push_back(logical_dispatch_core);
+            }
+        }
+    }
+
+    CoreCoord get_next_available_dispatch_core(chip_id_t device_id) {
```

Comment:
What is an "available_dispatch_core". Either needs better func name, or comments

Reply:
dispatch_core_manager first pulls all dispatch cores for a given device from the core_descriptor and then saves that into `std::unordered_map<chip_id_t, std::list<CoreCoord>> available_dispatch_cores_by_device` which is a list of unassigned dispatch cores per device id

when we try to get a specific dispatch core (e.g. the issue_queue_interface_core) we look into above map and pop a core from `available_dispatch_cores_by_device.at(device_id)` and mark that core as assigned.

Will add comments explaining this